File: mavkit-python/examples/monitor_telemetry.py
import asyncio
import os

import mavkit

CLEAR = "\x1b[2J\x1b[H"


def latest_value(sample):
    return sample.value if sample is not None else None


async def main():
    bind_addr = os.environ.get("MAVKIT_EXAMPLE_UDP_BIND", "0.0.0.0:14550")

    vehicle = await mavkit.Vehicle.connect_udp(bind_addr)
    telemetry = vehicle.telemetry()
    position_handle = telemetry.position().global_pos()
    groundspeed_handle = telemetry.position().groundspeed_mps()
    airspeed_handle = telemetry.position().airspeed_mps()
    climb_handle = telemetry.position().climb_rate_mps()
    heading_handle = telemetry.position().heading_deg()
    attitude_handle = telemetry.attitude().euler()
    battery_voltage_handle = telemetry.battery().voltage_v()
    battery_current_handle = telemetry.battery().current_a()
    battery_remaining_handle = telemetry.battery().remaining_pct()
    gps_quality_handle = telemetry.gps().quality()

    print("connected, streaming telemetry (Ctrl+C to stop)...\n")

    while True:
        position = (await position_handle.wait()).value
        attitude = latest_value(attitude_handle.latest())
        groundspeed = latest_value(groundspeed_handle.latest())
        airspeed = latest_value(airspeed_handle.latest())
        climb_rate = latest_value(climb_handle.latest())
        heading = latest_value(heading_handle.latest())
        battery_voltage = latest_value(battery_voltage_handle.latest())
        battery_current = latest_value(battery_current_handle.latest())
        battery_remaining = latest_value(battery_remaining_handle.latest())
        gps_quality = latest_value(gps_quality_handle.latest())

        print(CLEAR, end="")

        print("=== Position ===")
        print(f"  lat: {position.latitude_deg:>12.7f}°")
        print(f"  lon: {position.longitude_deg:>12.7f}°")
        print(f"  alt: {position.relative_alt_m:>8.1f} m")

        print("\n=== Movement ===")
        print(f"  speed:    {groundspeed or 0:>6.1f} m/s")
        print(f"  airspeed: {airspeed or 0:>6.1f} m/s")
        print(f"  climb:    {climb_rate or 0:>6.1f} m/s")
        print(f"  heading:  {heading or 0:>6.1f}°")

        print("\n=== Attitude ===")
        print(f"  roll:  {attitude.roll_deg if attitude else 0:>7.2f}°")
        print(f"  pitch: {attitude.pitch_deg if attitude else 0:>7.2f}°")
        print(f"  yaw:   {attitude.yaw_deg if attitude else 0:>7.2f}°")

        print("\n=== Battery ===")
        print(f"  voltage:   {battery_voltage or 0:>6.2f} V")
        print(f"  current:   {battery_current or 0:>6.2f} A")
        print(f"  remaining: {battery_remaining or 0:>3.0f}%")

        print("\n=== GPS ===")
        print(f"  fix:        {gps_quality.fix_type if gps_quality else None}")
        sats = (
            str(gps_quality.satellites)
            if gps_quality and gps_quality.satellites is not None
            else "-"
        )
        print(f"  satellites: {sats}")
        print(
            f"  hdop:       {gps_quality.hdop if gps_quality and gps_quality.hdop is not None else 0:.2f}"
        )


asyncio.run(main())

File: examples/monitor_telemetry.rs
use mavkit::{MetricSample, Vehicle};

fn latest_value<T>(sample: Option<MetricSample<T>>) -> Option<T> {
    sample.map(|sample| sample.value)
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let bind_addr =
        std::env::var("MAVKIT_EXAMPLE_UDP_BIND").unwrap_or_else(|_| "0.0.0.0:14550".to_string());

    let vehicle = Vehicle::connect_udp(&bind_addr).await?;
    let telemetry = vehicle.telemetry();
    let position_handle = telemetry.position().global();
    let groundspeed_handle = telemetry.position().groundspeed_mps();
    let airspeed_handle = telemetry.position().airspeed_mps();
    let climb_handle = telemetry.position().climb_rate_mps();
    let heading_handle = telemetry.position().heading_deg();
    let attitude_handle = telemetry.attitude().euler();
    let battery_voltage_handle = telemetry.battery().voltage_v();
    let battery_current_handle = telemetry.battery().current_a();
    let battery_remaining_handle = telemetry.battery().remaining_pct();
    let gps_quality_handle = telemetry.gps().quality();

    println!("connected, streaming telemetry (Ctrl+C to stop)...\n");

    loop {
        let position = position_handle.wait().await?.value;
        let attitude = latest_value(attitude_handle.latest());
        let groundspeed = latest_value(groundspeed_handle.latest());
        let airspeed = latest_value(airspeed_handle.latest());
        let climb_rate = latest_value(climb_handle.latest());
        let heading = latest_value(heading_handle.latest());
        let battery_voltage = latest_value(battery_voltage_handle.latest());
        let battery_current = latest_value(battery_current_handle.latest());
        let battery_remaining = latest_value(battery_remaining_handle.latest());
        let gps_quality = latest_value(gps_quality_handle.latest());

        print!("\x1b[2J\x1b[H"); // clear terminal

        println!("=== Position ===");
        println!("  lat: {:>12.7}°", position.latitude_deg);
        println!("  lon: {:>12.7}°", position.longitude_deg);
        println!("  alt: {:>8.1} m", position.relative_alt_m);

        println!("\n=== Movement ===");
        println!("  speed:    {:>6.1} m/s", groundspeed.unwrap_or_default());
        println!("  airspeed: {:>6.1} m/s", airspeed.unwrap_or_default());
        println!("  climb:    {:>6.1} m/s", climb_rate.unwrap_or_default());
        println!("  heading:  {:>6.1}°", heading.unwrap_or_default());

        println!("\n=== Attitude ===");
        println!(
            "  roll:  {:>7.2}°",
            attitude.as_ref().map_or(0.0, |value| value.roll_deg)
        );
        println!(
            "  pitch: {:>7.2}°",
            attitude.as_ref().map_or(0.0, |value| value.pitch_deg)
        );
        println!(
            "  yaw:   {:>7.2}°",
            attitude.as_ref().map_or(0.0, |value| value.yaw_deg)
        );

        println!("\n=== Battery ===");
        println!("  voltage: {:>6.2} V", battery_voltage.unwrap_or_default());
        println!("  current: {:>6.2} A", battery_current.unwrap_or_default());
        println!(
            "  remaining: {:>3.0}%",
            battery_remaining.unwrap_or_default()
        );

        println!("\n=== GPS ===");
        println!(
            "  fix:        {:?}",
            gps_quality.as_ref().map(|value| value.fix_type)
        );
        println!(
            "  satellites: {}",
            gps_quality
                .as_ref()
                .and_then(|value| value.satellites)
                .map_or("-".to_string(), |s| s.to_string())
        );
        println!(
            "  hdop:       {:.2}",
            gps_quality
                .as_ref()
                .and_then(|value| value.hdop)
                .unwrap_or(0.0)
        );
    }
}
